pyop2/petsc_base.py

```diff
@@ -627,6 +630,31 @@ def _init(self):
         else:
             self._init_block()
 
+    def _init_dense(self):
+        mat = PETSc.Mat()
+        rset, cset = self.sparsity.dsets
+        rlgmap = rset.unblocked_lgmap
+        clgmap = cset.unblocked_lgmap
+        mat.createDense(size=((self.nrows, None), (self.ncols, None)),
+                        bsize=1,
+                        comm=self.comm)
```

Comment:
I'm not sure this is right, since now your matrix will only support `MatSetValues` not `MatSetValuesBlocked` (and the decision we make as to whether to use one or the other depends on whether the maps are blocked, not the matrix).

Does it work if for assembling `VectorFunctionSpace` problems?

Reply:
I added a test on the firedrake PR, comparing `aij` and `dense` for a vector-valued problem. Works fine.

What do you think would be safer?

Reply:
Ah, now I understand, because you make it with `MatBlock` things, everything falls out right, because the matblocks have the right block size.